### PR TITLE
[topgen] Xbar autoconnect

### PR DESCRIPTION
**Pull request by eunchan**
Previously, the connection between the module, memory list and the xbar port shall be described manually in `top["inter_module"]["connect"]`.

Now topgen, after generating the Xbar ips, adds the connections if doesn't exist. It loops the node in every crossbars and check if it comes from the module or connects to the other xbars. Then it adds connection based on the information.

To make it easier and to not create double connection case, `add_intermodule_connect()` function is added. The function checks if the connection already exists and add if not.

### CODE REVIEW

File: util/topgen/intermodule.py [hunk 1]
```diff
@@ -69,6 +69,93 @@ def get_suffixes(ims: OrderedDict) -> (str, str):
     return ("_req", "_rsp")
 
 
+def add_intermodule_connection(obj, req_m, req_s, rsp_m, rsp_s):
```
Comment (msfschaffner) [inline util/topgen/intermodule.py:72]: nit: should we name these suffixes `_h` and `_d` instead?

File: util/topgen.py [hunk 2]
```diff
@@ -866,7 +866,11 @@ def main():
         generate_xbars(completecfg, out_path)
 
     # All IPs are generated. Connect phase now
-    elab_intermodule(completecfg)
+    # Find {memory, module} <-> {xbar} connections first.
+    im.autoconnect(completecfg)
```
Comment (tjaychen) [inline util/topgen.py:870]: does it make sense to call this fabric auto connect?
I'm guessing in the future, this will evolve into {memory, module, subsystem} <-> {xbar} right?

File: util/topgen/intermodule.py [hunk 3]
```diff
@@ -69,6 +69,93 @@ def get_suffixes(ims: OrderedDict) -> (str, str):
...
+                                           rsp_m=xbar["name"],
+                                           rsp_s="tl_%s" % port["name"])
+                continue  # len() == 1
+
+            # if not in module, memory, should be existed in top or ext field
+            module_key = "{}.tl_{}".format(xbar["name"], port["name"])
+            if module_key not in topcfg["inter_module"]["top"] + topcfg[
+                    "inter_module"]["external"]:
+                log.error("Inter-module key {} is not in module, memory list. "
+                          "Shall be in top or ext".format(module_key))
```
Comment (msfschaffner) [inline util/topgen/intermodule.py:156]: nit: maybe rephrase as
`The inter-module key {} can not be found in the module / memory lists, nor in the top / ext lists.`

File: util/topgen/intermodule.py [hunk 4]
```diff
@@ -69,6 +69,93 @@ def get_suffixes(ims: OrderedDict) -> (str, str):
...
+    # Add new key and connect
+    connect[req_key] = [rsp_key]
+
+
+def autoconnect(topcfg: OrderedDict):
+    """Matching the connection based on the naming rule
+    between {memory, module} <-> Xbar.
+    """
+
+    instances = topcfg["module"] + topcfg["memory"]
```
Comment (tjaychen) [inline util/topgen/intermodule.py:104]: this is not for this PR.  Do you think we should create "shell" hjson entries for things like ibex, rv_dm etc? Then maybe long term we can hook those into auto-connect also.

File: util/topgen/intermodule.py [hunk 5]
```diff
@@ -69,6 +69,93 @@ def get_suffixes(ims: OrderedDict) -> (str, str):
...
+                    continue
+                add_intermodule_connection(obj=topcfg,
+                                           req_m=xbar["name"],
+                                           req_s="tl_%s" % port["name"],
+                                           rsp_m=port["name"],
+                                           rsp_s="tl_%s" % xbar["name"])
+
+            # check if name exists in ["module", "memory"]
+            ips = [x for x in instances if x["name"] == port["name"]]
+            if len(ips) == 1:
```
Comment (tjaychen) [inline util/topgen/intermodule.py:121]: should this be an error if length > 1?

File: util/topgen/intermodule.py (hunk 1 above)
Comment (tjaychen) [reply util/topgen/intermodule.py:72]: +1

File: util/topgen.py (hunk 2 above)
Comment (eunchan) [reply util/topgen.py:870]: Ah.. I should've consider the subsystem too... Let me think

File: util/topgen/intermodule.py (hunk 1 above)
Comment (eunchan) [reply util/topgen/intermodule.py:72]: You mean not `_h2d` `_d2h` but `_h`, `_d` for tlul signal? New implementation just does `_req` `_rsp` for TL-UL port as same manner as other intermodule signal.

File: util/topgen/intermodule.py (hunk 4 above)
Comment (eunchan) [reply util/topgen/intermodule.py:104]: yeah. that is always on my todo list.. :)

File: util/topgen/intermodule.py (hunk 3 above)
Comment (tjaychen) [reply util/topgen/intermodule.py:156]: why don't we just say `can not be found in module, memory, top or ext lists`?

Comment (eunchan) [reply util/topgen/intermodule.py:156]: Looks good. Let me revise

File: util/topgen/intermodule.py (hunk 1 above)
Comment (tjaychen) [reply util/topgen/intermodule.py:72]: o no, i think @msfschaffner meant the input variable names, instead of `req_m`, we use `req_h`.  `rsp_d` instead of `rsp_s`

File: util/topgen/intermodule.py (hunk 5 above)
Comment (eunchan) [reply util/topgen/intermodule.py]: Thanks, revised slightly . New code has smaller indentations :)

File: util/topgen/intermodule.py (hunk 1 above)
Comment (eunchan) [reply util/topgen/intermodule.py:72]: it is "module" and "signal". Not the overloaded terms :)

Comment (tjaychen) [reply util/topgen/intermodule.py:72]: o dang, you're right.
Could you just add a comment in the definition header say that's what it is? :)
I'm sure i won't be the last to ask.

Comment (eunchan) [reply util/topgen/intermodule.py]: Revised

Comment (msfschaffner) [reply util/topgen/intermodule.py]: ahh got it. sorry for the false alarm :)
